aten/src/ATen/native/Convolution.cpp

```diff
@@ -1697,22 +1697,28 @@ std::tuple<Tensor, Tensor, Tensor> convolution_backward(
       std::tie(backend_grad_input, backend_grad_weight, backend_grad_bias) =
         miopen_convolution_backward_stub(
           input.device().type(),
-          input.contiguous(backend_memory_format), grad_output, weight, params.padding, params.stride,
+          // Only make input contiguous when it is necessary for the backwards computation
+          output_mask[1] ? input.contiguous(backend_memory_format) : input,
+          grad_output, weight, params.padding, params.stride,
           params.dilation, params.groups, params.benchmark, params.deterministic, output_mask);
       break;
     case ConvBackend::MiopenDepthwise:
       std::tie(backend_grad_input, backend_grad_weight, backend_grad_bias) =
           miopen_depthwise_convolution_backward_stub(
             input.device().type(),
-            input.contiguous(backend_memory_format), grad_output, weight, params.padding, params.stride,
+            // Only make input contiguous when it is necessary for the backwards computation
+            output_mask[1] ? input.contiguous(backend_memory_format) : input,
+            grad_output, weight, params.padding, params.stride,
```

Comment:
Do we need to set `weight.contiguous(backend_memory_format)` when `grad_input` is being computed (i.e. `output_mask[0]` is set) for the depthwise case?

Consider the case:
* `input` is channels_last
* `weight` is **not** channels_last
* `grad_output` is **not** channels_last
* `grad_input` is being computed, which uses `weight` & `grad_output`

We want to use channels_last due to the input. However, the decision to use channels_last is currently based on weight and grad_output:
https://github.com/pytorch/pytorch/blob/2dbbb1a921e23b7e492a1ad63f9420ad49d15713/aten/src/ATen/native/miopen/Conv_miopen.cpp#L1183-L1185

So I think the logic will wrongly choose not to use channels_last for this case, unless we set `weight.contiguous(backend_memory_format)` before the backward call.

Or am I reading this wrong?

Reply:
What does it mean for the weight to be channels last? If an input is channels last then it is has NCHW shape but NHWC strides. If a weight has channels last, then it has OIHW shape and ???? strides -- is ???? "OHWI"?

Reply:
My understanding is that channels_last is *technically* not a property of the tensor, but more an indicator of how to set memory_format for the result. From the comment at the top of `MemoryFormat.h`:
https://github.com/pytorch/pytorch/blob/4868907cf31eac0255c1e0cb205db6d70c70fcd8/c10/core/MemoryFormat.h#L9-L12

`weight` being channels_last means that something like this has been done, which I believe is one way the user can specify that channels_last should be used for the result.
```python
import torch

m = torch.nn.Conv2d(3, 3, 3)
m.to(memory_format=torch.channels_last)
print(m.weight.is_contiguous(memory_format=torch.channels_last))  # True
```

I think if the user does this, they'll expect the result to be in channels_last. Also, based on the channels_last semantics, passing an input in channels_last should also produce a result in channels_last. From the [memory_format tutorial](https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html): `Convolution modules, unlike binary p-wise operator, have channels last as the dominating memory format. IFF all inputs are in contiguous memory format, the operator produces output in contiguous memory format. Otherwise, output wil be in channels last memroy format.`

I believe @ngimel knows more about memory_format so she can correct me if I'm wrong here.

Reply:
Related: #70125 - it seems to be the case that memory_format support is broken on ROCm right now. Ideally we should fix this and stop skipping `test_memory_format`